### PR TITLE
SALTO-5439: [Okta] Extract AppGroupAssignments to their own instances 

### PR DESCRIPTION
**Pull request by shir-reifenberg**
Extract App-Group assignment to standalone fields

---

The motivation for this change is - 
1. As groups are not necessarily aligned between environments, users might want to deploy application without group assignments 
2. For provisioned apps, group assignments might include additional app-specific info that we don't fetch today

Changes - 
- Create new AppGroupAssignment instances
- Modify the app group validator to work with the new instances
- add dependency changer to verify we always deploy modifications of applications before any ApplicationGroupAssignment changes, because we changes to assignments cannot be done to inactive apps, we should make status changes before deployment of ApplicationGroupAssignment

---
_Release Notes_: 

_Okta_adapter_:
- Application group assignments will be moved to their own instance.

---
_User Notifications_: 

_Okta_adapter_:
- Application group assignment configurations will no longer be part of Application instances, and will have their own instances.

### CODE REVIEW

File: packages/okta-adapter/src/config.ts [hunk 1]
```diff
@@ -434,6 +434,42 @@ const DEFAULT_TYPE_CUSTOMIZATIONS: OktaSwaggerApiConfig['types'] = {
...
+      fieldTypeOverrides: [{ fieldName: 'profile', fieldType: 'map<unknown>' }],
+      serviceUrl: '/admin/app/{_parent.0.name}/instance/{_parent.0.id}/#tab-assignments',
+    },
+    deployRequests: {
+      add: {
+        url: '/api/v1/apps/{appId}/groups/{groupId}',
+        method: 'put',
+        urlParamsToFields: {
+          appId: '_parent.0.id',
+          groupId: 'id',
```
Comment (idoamit8) [inline packages/okta-adapter/src/config.ts:451]: nit: we can remove groupId and write {id} in the url (same goes to modify and remove)
``` suggestion
       url: '/api/v1/apps/{appId}/groups/{id}',
        method: 'put',
        urlParamsToFields: {
          appId: '_parent.0.id',
```

File: packages/okta-adapter/src/dependency_changers/app_group_assignment_to_app.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    change => getChangeData(change.change).elemID.typeName === APP_GROUP_ASSIGNMENT_TYPE_NAME,
+  )
+
+  if (_.isEmpty(appModificationChanges) || _.isEmpty(appGroupChanges)) {
+    return []
+  }
+
+  return appGroupChanges
+    .map(appGroupChange => {
+      const parentApp = getParent(getChangeData(appGroupChange.change))
```
Comment (idoamit8) [inline packages/okta-adapter/src/dependency_changers/app_group_assignment_to_app.ts:61]: `getParent` can throw. Please add try-catch or use `hasValidParent`

File: packages/okta-adapter/src/change_validators/app_group_assignments.ts [hunk 3]
```diff
@@ -13,59 +13,44 @@
...
-  if (!isArrayOfReferencesOrUndefined(beforeAssignments) || !isArrayOfReferencesOrUndefined(afterAssignments)) {
+const isParentAppInStatusInactive = (appGroupInstance: InstanceElement): boolean => {
+  try {
+    const parent = getParent(appGroupInstance)
+    return parent.elemID.typeName === APPLICATION_TYPE_NAME && parent.value.status === INACTIVE_STATUS
+  } catch (err) {
+    log.error(
+      'Failed to get parent app, skipping appGroupAssignmentValidator for %s',
+      appGroupInstance.elemID.getFullName(),
+    )
```
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/okta-adapter/src/change_validators/app_group_assignments.ts:37]: IMO if I use the parent and not just validating it's exists, there's no point calling `hasValidParent`

[the remark this answers is not in the document]
Comment (idoamit8) [reply packages/okta-adapter/src/change_validators/app_group_assignments.ts:37]: I think that we we should change the `getParent` to not throw, and thats why I suggested it (So we can later modify the `getParent` func and the `hasValidParent` func. 
But both ways are good.

**Pull request merged**